include/seqan3/alphabet/detail/alphabet_base.hpp

```diff
@@ -0,0 +1,237 @@
+ *
+ *   * `static std::array<char_type, value_size> constexpr rank_to_char` that defines for every possible rank value
+ *     the corresponding char value.
+ *   * `static std::array<rank_type, 256> constexpr char_to_rank` that defines for every possible character value the
+ *     corresponding rank value.
+ *
+ * TODO add snippet
+ */
+template <typename derived_type, size_t size, typename char_t = char>
+class alphabet_base
+{
+public:
+    //TODO: this should be > 1 and there should be a different specialisation for size 1
+    static_assert(size > 0, "It does not make sense to use the base class for alphabets of size <= 1.");
+
+    /*!\name Member types
+     * \{
+     */
+    //!\brief The type of the alphabet when converted to char (e.g. via to_char()).
+    using char_type = char_t;
+    //!\brief The type of the alphabet when represented as a number (e.g. via to_rank()).
+    using rank_type = detail::min_viable_uint_t<size - 1>;
+    //!\}
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr alphabet_base() = default;
+    constexpr alphabet_base(alphabet_base const &) = default;
+    constexpr alphabet_base(alphabet_base &&) = default;
+    constexpr alphabet_base & operator=(alphabet_base const &) = default;
+    constexpr alphabet_base & operator=(alphabet_base &&) = default;
+    //!\}
+
+    /*!\name Read functions
+     * \{
+     */
+    /*!\brief Return the letter as a character of char_type.
+     *
+     * \details
+     *
+     * Satisfies the seqan3::alphabet_concept::to_char() requirement via the seqan3::to_char() wrapper.
+     *
+     * \par Complexity
+     *
+     * Constant.
+     *
+     * \par Exceptions
+     *
+     * Guaranteed not to throw.
+     */
+    constexpr char_type to_char() const noexcept
+    //!\cond
+        requires !std::Same<char_type, void>
+    //!\endcond
+    {
+        return derived_type::rank_to_char[rank];
+    }
+
+    /*!\brief Return the letter's numeric value or rank in the alphabet.
+     *
+     * \details
+     *
+     * Satisfies the seqan3::semi_alphabet_concept::to_rank() requirement via the to_rank() wrapper.
+     *
+     * \par Complexity
+     *
+     * Constant.
+     *
+     * \par Exceptions
+     *
+     * Guaranteed not to throw.
+     */
+    constexpr rank_type to_rank() const noexcept
+    {
+        return rank;
+    }
+    //!\}
+
+    /*!\name Write functions
+     * \{
+     */
+    /*!\brief Assign from a character.
+     *
+     * \details
+     *
+     * Satisfies the seqan3::alphabet_concept::assign_char() requirement via the seqan3::assign_char() wrapper.
+     *
+     * \par Complexity
+     *
+     * Constant.
+     *
+     * \par Exceptions
+     *
+     * Guaranteed not to throw.
+     */
+    constexpr derived_type & assign_char(std::conditional_t<std::Same<char_type, void>, char, char_type> const c) noexcept
+    //!\cond
+        requires !std::Same<char_type, void>
```

Comment:
If you explicitly require that the type is not void, why do you need the `std::conditional` (or the other way around)

Reply:
Because if the parameter type is void, the program becomes ill-formed (even though the function would never be chosen due to the concept check). I know, it's not pretty.